larpandoradlcontent/LArTwoDReco/DLThreeDClusterSplittingAlgorithm.cc

```diff
@@ -0,0 +1,1350 @@
+                     if (iCurrent == iTest)
+                         continue;
+
+                     const float endpointSep((splitPositions.at(iCurrent) - testPositions.at(iTest)).GetMagnitude());
+                     float openingAngle(splitDirections.at(iCurrent).GetOpeningAngle(testDirections.at(iTest)));
+                     openingAngle *= (180.f / 3.14);
+                                            
+                     if ((endpointSep < m_collinearMaxSep) && 
+                         ((openingAngle < m_collinearMaxOpeningAngle) || (openingAngle > (180.f - m_collinearMaxOpeningAngle))))
+                     {
+                         collinear = true;
+                     }
+                 }
+
+                 if (collinear)
+                     rejected.push_back(iCurrent);
+            }
+        }
+
+        // Finally, add split positions
+        for (int i = 0; i < 2; ++i)
+        {
+            if (std::find(rejected.begin(), rejected.end(), i) == rejected.end())
+                trueSplitPositions.push_back(splitPositions.at(i));
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DLThreeDClusterSplittingAlgorithm::FillTree(const CartesianPointVector &splitPositions, const MCParticleToHitListMap &mainHitListMap, 
+    const TwoDSlidingFitResult &clusterFit, const Features &features)
+{
+    const bool isContaminated(!splitPositions.empty());
+
+    // Work out who truly owns this cluster
+    int highestNHits(-1), matchedPDG(-1);
+    for (const auto &entry : mainHitListMap)
+    {
+        if (static_cast<int>(entry.second.size()) > highestNHits)
+        {
+            highestNHits = entry.second.size();
+            matchedPDG = entry.first->GetParticleId();
+        }
+    }
+
+    // Don't want to to train on very contaminated tracks
+    int nContaminants(0);
+    for (auto &entry : mainHitListMap)
+        if (entry.second.size() >= m_reducedMinTargetMCHits)
+            ++nContaminants;
+
+    // Determine l of splitting positions
+    FloatVector vertexL;
+    if (isContaminated)
+    {
+        for (const CartesianVector &splitPosition : splitPositions)
+        {
+            float thisL(0.f), thisT(0.f);
+            clusterFit.GetLocalPosition(splitPosition, thisL, thisT);
+            vertexL.push_back(thisL);
+        }
+    }
+
+    // Fill tree
+    FloatVector energy(*features.at("Energy").GetSequence()), hitWidth(*features.at("HitWidth").GetSequence()), 
+        gapSep(*features.at("GapSep").GetSequence()), eventHitSep(*features.at("EventHitSep").GetSequence()), 
+        clusterHitSep(*features.at("ClusterHitSep").GetSequence()), longitudinal(*features.at("Longitudinal").GetSequence()),
+        transverse(*features.at("Transverse").GetSequence()), theta(*features.at("Theta").GetSequence()),
+        secVertex(*features.at("SecVertex").GetSequence());
```

Comment:
I think this relates to an earlier comment about returning a const reference - I think if you do that, all of this can go and you can just pass the result of the function call to SetTreeVariable below (with suitable & for the vectors).